book/embeds.md

```diff
@@ -102,26 +91,75 @@ def request(url, top_level_url, payload=None):
     # ...
     while True:
         line = response.readline().decode("utf8")
-        # ...    
-```
-
-Then when processing the response body, check for the `content-type` header,
-which will tell the browser how to decode that part of the HTTP response. I
-discussed this header briefly in [Chapter 1](/http.html#the-servers-response), 
-where I noted that HTML web page responses have a value of `text/html` for this
-header. This value is a [MIME type][mime-type], which stands for Multipurpose
-Internet Mail Extensions, and was originally intended for enumerating all of the
-acceptable data formats for email attachments.^[Most email these days is
-actually HTML, and is encoded with the `text/html` MIME type. Gmail, for
-example, by default uses this format, but can be put in a "plain text mode"
-that encodes the email in `text/plain`.] We've actually encountered two more
-content types already: `text/css` and `application/javascript`, but since
-it assumed both were in `utf8` there was no need to differentiate in the
-code.^[That's not a correct thing to do in a real browser, and alternate
-character sets are an exercise in chapter 1.]
+        # ...
+    # ...
+```
+
+Note that I _didn't_ add a `decode` call when we read the body; that's
+because the body actually might be binary data, and we want to return
+that binary data directly to the browser. Now, every existing call to
+`request`, which wants textual data, needs to `decode` the response.
+For example, in `load`, you'll want to do something like this:
+
+``` {.python replace=Tab/Frame}
+class Tab:
+    def load(self, url, body=None):
+        # ...
+        headers, body = request(url, self.url, payload=body)
+        body = body.decode("utf8")
+        # ...
+```
+
+Make sure to make this change everywhere in your browser that you call
+`request`, including inside `XMLHttpRequest_send` and in several other
+places in `request`. When we download images, however, we _won't_ call
+`decode`, and just use the binary data directly:
+
+``` {.python replace=tab/frame}
+def download_image(image_src, tab):
+    image_url = resolve_url(image_src, tab.url)
+    if not tab.allowed_request(image_url):
+        print("Blocked image", link, "due to CSP")
+        return
+    try:
+        header, body = request(image_url, tab.url)
+        return PIL.Image.open(io.BytesIO(body))
+    except:
+        print("Failed to load image: " + image_url)
+```
+
+This method wraps the image bytes as a `BytesIO` object, and then
+passes that to `open` to turn it into an Pillow `Image` object.
+
+::: {.todo}
+
+Chris, do we actually need MIME types? It seems kind of dangerous the
```

Comment:
I guess we don't need it after your proposed refactoring.

Reply:
We could remove it or move it to a go-further block.

Reply:
I'll make a go-further block out of it.